From 06ac433892fac315d83d7ead5e240cd2d1cb2f53 Mon Sep 17 00:00:00 2001
From: Felix Kuehling <Felix.Kuehling@amd.com>
Date: Thu, 6 Oct 2016 15:52:07 -0400
Subject: [PATCH 1678/1722] drm/amdgpu: Fix parsing of power profiles

Need to copy the NULL string terminator. Also return failure, if too
many parameters are written by user mode.

Change-Id: I1e74989e9db1a1c004ddab7c1682fb2411fe6ee6
Signed-off-by: Felix Kuehling <Felix.Kuehling@amd.com>
---
 drivers/gpu/drm/amd/amdgpu/amdgpu_pm.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_pm.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_pm.c
index 68cdde5..f123631 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_pm.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_pm.c
@@ -650,12 +650,12 @@ static ssize_t amdgpu_set_pp_power_profile(struct device *dev,
 		return count;
 	}
 
-	if (count >= 128) {
+	if (count+1 >= 128) {
 		count = -EINVAL;
 		goto fail;
 	}
 
-	memcpy(buf_cpy, buf, count);
+	memcpy(buf_cpy, buf, count+1);
 	tmp_str = buf_cpy;
 
 	while (tmp_str[0]) {
@@ -685,7 +685,8 @@ static ssize_t amdgpu_set_pp_power_profile(struct device *dev,
 			request->down_hyst = (uint8_t)value;
 			break;
 		default:
-			break;
+			count = -EINVAL;
+			goto fail;
 		}
 
 		loop++;
-- 
2.7.4

